### ISSUE TITLE
HunyuanOCR微调loss异常

### ISSUE THREAD

**Issue by wusijie123** [modelscope/ms-swift#7127]
使用最新的HunyuanOCR全参数微调，初始loss为4以上，日志中未见相关异常信息，而token_acc为0.3左右，也不会是模型初始化了，排查发现模板和hf的拼接有所出入，hf的图片拼接<｜hy_place▁holder▁no▁100｜>，<｜hy_place▁holder▁no▁101｜>，分别对应120118，120119 token，而log打印的样本没有，在swift/llm/template/template/tencent.py对齐后loss仍然存在异常

`{'loss': 4.42422962, 'grad_norm': 45.37061691, 'learning_rate': 4e-08, 'token_acc': 0.28018576, 'epoch': 0.0, 'global_step/max_steps': '1/5368', 'percentage': '0.02%', 'elapsed_time': '7s', 'remaining_time': '11h 31m 19s', 'memory(GiB)': 4.73, 'train_speed(iter/s)': 0.12939}`

启动脚本：
torchrun \
    --nproc_per_node $NPUS_PER_NODE \
    --nnodes $NNODES \
    --node_rank $NODE_RANK \
    --master_addr $MASTER_ADDR \
    --master_port $MASTER_PORT \
    swift/cli/sft.py \
    --model_type hunyuan_ocr \
    --model $BASE_MODEL_DIR \
    --dataset $DATA_PATH \
    $resume_param \
    --train_type full \
    --torch_dtype bfloat16 \
    --num_train_epochs 2 \
    --per_device_train_batch_size 1 \
    --per_device_eval_batch_size 1 \
    --learning_rate 1e-7 \
    --max_pixels 1048576 \
    --gradient_accumulation_steps 1 \
    --save_steps $save_steps \
    --eval_strategy no \
    --save_total_limit 2 \
    --logging_steps 10 \
    --max_length 4096 \
    --output_dir /home/vl-finetune/$MODEL_ID \
    --warmup_ratio 0.05 \
    --lazy_tokenize true \
    --ddp_backend hccl \
    --freeze_aligner false \
    --lr_scheduler_type=cosine \
    --freeze_vit false \
    --deepspeed zero2 \
    --dataloader_num_workers 8

<img width="2546" height="384" alt="Image" src="https://github.com/user-attachments/assets/94963d93-69e2-416e-9ef7-476b3f417ceb" />

**Comment by Jintao-Huang** [comment]
稍等 我来修复

**Comment by Jintao-Huang** [comment]
我对过input_ids是没有问题的

**Comment by Jintao-Huang** [comment]
https://github.com/modelscope/ms-swift/pull/7143

这里修复了batch_size>1

**Comment by wusijie123** [comment]
使用最新的example脚本，loss还是有异常，input_id和之前也是一样的
`The tokenizer has new PAD/BOS/EOS tokens that differ from the model config and generation config. The model config and generation config were aligned accordingly, being updated with the tokenizer's values. Updated tokens: {'eos_token_id': 120007}.
Train:   0%|                                                                                                                                                                                           | 0/62 [00:00<?, ?it/s][INFO:swift] use_logits_to_keep: False
{'loss': 8.86149025, 'grad_norm': 18.00219536, 'learning_rate': 2.5e-05, 'token_acc': 0.06504065, 'epoch': 0.02, 'global_step/max_steps': '1/62', 'percentage': '1.61%', 'elapsed_time': '7s', 'remaining_time': '8m 7s', 'memory(GiB)': 4.81, 'train_speed(iter/s)': 0.12523}
{'loss': 8.7316494, 'grad_norm': 17.23290634, 'learning_rate': 9.993e-05, 'token_acc': 0.07068607, 'epoch': 0.08, 'global_step/max_steps': '5/62', 'percentage': '8.06%', 'elapsed_time': '18s', 'remaining_time': '3m 36s', 'memory(GiB)': 4.87, 'train_speed(iter/s)': 0.263319}
{'loss': 7.03114243, 'grad_norm': 5.50963402, 'learning_rate': 9.738e-05, 'token_acc': 0.09538462, 'epoch': 0.16, 'global_step/max_steps': '10/62', 'percentage': '16.13%', 'elapsed_time': '31s', 'remaining_time': '2m 45s', 'memory(GiB)': 4.87, 'train_speed(iter/s)': 0.313901}
{'loss': 6.32034149, 'grad_norm': 12.17352867, 'learning_rate': 9.138e-05, 'token_acc': 0.17752443, 'epoch': 0.24, 'global_step/max_steps': '15/62', 'percentage': '24.19%', 'elapsed_time': '50s', 'remaining_time': '2m 37s', 'memory(GiB)': 4.88, 'train_speed(iter/s)': 0.297648}
{'loss': 6.68428268, 'grad_norm': 4.91635466, 'learning_rate': 8.237e-05, 'token_acc': 0.169967, 'epoch': 0.32, 'global_step/max_steps': '20/62', 'percentage': '32.26%', 'elapsed_time': '1m 2s', 'remaining_time': '2m 10s', 'memory(GiB)': 4.88, 'train_speed(iter/s)': 0.321278}
{'loss': 6.3566906, 'grad_norm': 4.97650003, 'learning_rate': 7.099e-05, 'token_acc': 0.1703125, 'epoch': 0.4, 'global_step/max_steps': '25/62', 'percentage': '40.32%', 'elapsed_time': '1m 17s', 'remaining_time': '1m 54s', 'memory(GiB)': 4.88, 'train_speed(iter/s)': 0.321977}
{'loss': 6.42645493, 'grad_norm': 5.99621916, 'learning_rate': 5.809e-05, 'token_acc': 0.16524702, 'epoch': 0.48, 'global_step/max_steps': '30/62', 'percentage': '48.39%', 'elapsed_time': '1m 32s', 'remaining_time': '1m 38s', 'memory(GiB)': 4.88, 'train_speed(iter/s)': 0.324114}
{'loss': 6.09507675, 'grad_norm': 4.81372595, 'learning_rate': 4.459e-05, 'token_acc': 0.18630573, 'epoch': 0.56, 'global_step/max_steps': '35/62', 'percentage': '56.45%', 'elapsed_time': '1m 48s', 'remaining_time': '1m 23s', 'memory(GiB)': 4.88, 'train_speed(iter/s)': 0.323728}
{'loss': 6.07139587, 'grad_norm': 4.15322447, 'learning_rate': 3.149e-05, 'token_acc': 0.16981132, 'epoch': 0.65, 'global_step/max_steps': '40/62', 'percentage': '64.52%', 'elapsed_time': '2m 3s', 'remaining_time': '1m 7s', 'memory(GiB)': 4.88, 'train_speed(iter/s)': 0.324193}`
可以分享一下你实测log的input_ids吗

**Comment by Jintao-Huang** [comment]
试试这个数据集

AI-ModelScope/LaTeX_OCR#5000

**Comment by wusijie123** [comment]
> 试试这个数据集
> 
> AI-ModelScope/LaTeX_OCR#5000

感谢，这个是没问题的

**Comment by slin000111** [comment]
> 使用最新的HunyuanOCR全参数微调，初始loss为4以上，日志中未见相关异常信息，而token_acc为0.3左右，也不会是模型初始化了，排查发现模板和hf的拼接有所出入，hf的图片拼接<｜hy_place▁holder▁no▁100｜>，<｜hy_place▁holder▁no▁101｜>，分别对应120118，120119 token，而log打印的样本没有，在swift/llm/template/template/tencent.py对齐后loss仍然存在异常
> 
> `{'loss': 4.42422962, 'grad_norm': 45.37061691, 'learning_rate': 4e-08, 'token_acc': 0.28018576, 'epoch': 0.0, 'global_step/max_steps': '1/5368', 'percentage': '0.02%', 'elapsed_time': '7s', 'remaining_time': '11h 31m 19s', 'memory(GiB)': 4.73, 'train_speed(iter/s)': 0.12939}`
> 
> 启动脚本： torchrun --nproc_per_node $NPUS_PER_NODE --nnodes $NNODES --node_rank $NODE_RANK --master_addr $MASTER_ADDR --master_port $MASTER_PORT swift/cli/sft.py --model_type hunyuan_ocr --model $BASE_MODEL_DIR --dataset $DATA_PATH $resume_param --train_type full --torch_dtype bfloat16 --num_train_epochs 2 --per_device_train_batch_size 1 --per_device_eval_batch_size 1 --learning_rate 1e-7 --max_pixels 1048576 --gradient_accumulation_steps 1 --save_steps $save_steps --eval_strategy no --save_total_limit 2 --logging_steps 10 --max_length 4096 --output_dir /home/vl-finetune/$MODEL_ID --warmup_ratio 0.05 --lazy_tokenize true --ddp_backend hccl --freeze_aligner false --lr_scheduler_type=cosine --freeze_vit false --deepspeed zero2 --dataloader_num_workers 8
> 
> <img alt="Image" width="2000" height="384" src="https://private-user-images.githubusercontent.com/129249910/528565551-94963d93-69e2-416e-9ef7-476b3f417ceb.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njg3ODkxMTcsIm5iZiI6MTc2ODc4ODgxNywicGF0aCI6Ii8xMjkyNDk5MTAvNTI4NTY1NTUxLTk0OTYzZDkzLTY5ZTItNDE2ZS05ZWY3LTQ3NmIzZjQxN2NlYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDExOVQwMjEzMzdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMGM5MmExYTBmMmRkNTE2OGI4N2E2ODM0ZTA3MDIwYzBhNjc5ZWI3OTAxNTEwMTI4NTVlZGRkZTgxMzcxYTZlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.0vdfyf3yTk3K6Tinpjt5ctL4SJKyG7UnoAhwateqhBg">

再试试看，https://github.com/modelscope/ms-swift/pull/7803
